File: examples/echo_server.cc
// iox — unified async IO for Linux
// examples/echo_server.cc — Each accepted connection gets a detached io::loop session that echoes
#include <array>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <memory>
#include <string_view>

#include <iox/core/exec.h>
#include <iox/compose/loop.h>
#include <iox/compose/write_all.h>
#include <iox/compose/detach.h>
#include <iox/net/tcp.h>
#include <iox/ops.h>

int main(int argc, char** argv) {
    const std::uint16_t port = argc > 1 ? static_cast<std::uint16_t>(
                                              std::strtoul(argv[1], nullptr, 10))
                                        : 7777;

    iox::io_context context;
    auto listener = iox::net::tcp::acceptor::listen(*iox::net::endpoint::ipv4_any(port));
    if (!listener) {
        std::fprintf(stderr, "listen: %s\n", listener.error().message().c_str());
        return 1;
    }
    std::printf("echoing on 127.0.0.1:%u (Ctrl-C to stop)\n", port);

    struct session {
        iox::net::tcp::socket socket;
        std::unique_ptr<std::byte[]> buffer = std::make_unique<std::byte[]>(4096);

        explicit session(iox::net::tcp::socket sock) : socket(std::move(sock)) {}
    };

    auto serve_one = [&](iox::net::tcp::socket sock) {
        auto* session_ptr = new session{std::move(sock)};
        auto body = iox::io::loop(context, [session_ptr, &context, eof = false]() mutable {
            return iox::io::read(context, session_ptr->socket, iox::wbytes{session_ptr->buffer.get(), 4096})
                 | iox::exec::let_value([session_ptr, &context, &eof](std::size_t count) {
                       eof = (count == 0);
                       return iox::io::write_all(context, session_ptr->socket,
                                                 iox::rbytes{session_ptr->buffer.get(), count});
                   })
                 | iox::exec::then([&eof]() { return eof; });
        });
        iox::exec::detach(std::move(body) | iox::exec::upon_error([](auto&&) {})
                                 | iox::exec::then([session_ptr] { delete session_ptr; }));
    };

    auto accept_loop = iox::io::loop(context, [&]() {
        return iox::io::accept(context, *listener) | iox::exec::then(serve_one);
    });

  // The accept loop never yields true; run it detached and serve forever.
    iox::exec::detach(std::move(accept_loop));
    context.run();
    return 0;
}

File: examples/python/echo_server.py
#!/usr/bin/env python3
"""echo_server — a TCP echo server through iox (Python twin of
examples/echo_server.cc).

One Context, one thread: accept() parks on io_uring with the GIL released,
each connection is echoed read→write_all until EOF, then the server goes
back to accepting. Sequential by design — the iox context is
single-threaded; serve many clients concurrently by giving each thread
its own Context.

    python3 echo_server.py [PORT]
"""
import sys

import iox


def serve_connection(context: iox.Context, conn) -> None:
    peer_gone = False
    while not peer_gone:
        data = conn.recv(4096)
        if not data:  # EOF
            break
        conn.send(data)
    conn.close()


def main() -> int:
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 9000

    context = iox.Context()
    listener = iox.listen(context, "127.0.0.1", port)
    print(f"echoing on 127.0.0.1:{port} (ctrl-c to stop)", flush=True)
    while True:
        conn = listener.accept()  # blocks with the GIL released
        serve_connection(context, conn)


if __name__ == "__main__":
    try:
        sys.exit(main())
    except KeyboardInterrupt:
        sys.exit(0)
